armada/src/main.rs

```diff
@@ -1,56 +1,88 @@
 mod args;
 mod ranges;
 mod run_variants;
+use std::collections::HashMap;
+use csv::Writer;
+use std::error::Error;
 
-use std::net::{
+use std::{net::{
     IpAddr,
     Ipv4Addr,
-    Ipv6Addr,
-};
+    Ipv6Addr, SocketAddr,
+}, io::ErrorKind};
 
 use armada_lib::Armada;
 
 use crate::args::ArmadaCLIConfig;
 
+const DEFAULT_REPORT_NAME: &str = "scan_results.csv";
+
+
 #[tokio::main]
 async fn main() {
     let ArmadaCLIConfig {
         targets,
-        ports,
         quiet_mode,
+        ports,
         rate_limit,
         listening_port,
         retries,
         timeout,
         source_ips,
-        stream_results
+        stream_results,
+        generate_report
     } = args::get_armada_cli_config();
 
     let armada = Armada::new(listening_port);
 
     let (source_ipv4, source_ipv6) = split_and_enforce_source_ips(source_ips).await;
 
-    let mut syn_scan_results = if quiet_mode {
+    let syn_scan_results = if quiet_mode {
         use run_variants::QuietArmada;
 
         armada
-            .run_quiet(targets, ports, source_ipv4, source_ipv6, retries, timeout, rate_limit, stream_results)
+            .run_quiet(targets, ports, source_ipv4, source_ipv6, retries, timeout, rate_limit, stream_results, generate_report)
             .await
     } else {
         use run_variants::ProgressArmada;
 
         armada
-            .run_with_stats(targets, ports, source_ipv4, source_ipv6, retries, timeout, rate_limit, stream_results)
+            .run_with_stats(targets, ports, source_ipv4, source_ipv6, retries, timeout, rate_limit, stream_results, generate_report)
             .await
     };
 
-    if !stream_results {
-        syn_scan_results.sort();
 
-        syn_scan_results.into_iter().for_each(|remote| {
+    if !stream_results {
+        let scan_results = &syn_scan_results;
```

Comment:
@d0nutptr I'm not sure of the most performant way to do this. Ideally we would avoid unnecessary allocations, but into_iter() is causing issues with borrowing for anything else that may reference it later
 
I think cloning just in the scope of report generating should be ok though?